#163445913 implementation logic feedback

openwell opened the pull request:
#### What does this PR do?
- This PR integrates improved logic for better performance 
#### Description of Task to be completed?
-  Accessing of API end points
#### How should this be manually tested?
- Pull this branch [ch-implementation-logic-feedback-163445913](https://github.com/openwell/questioner/tree/ch-implementation-logic-feedback-163445913) off this repository
- On your command line, run  `npm install` to install all dependencies
- Run `npm database` to create the database tables and create an admin
- An admin token is generated if successfully and will be displayed in the terminal
- The token should be part of the headers as 'tokens': 'token-code'
- Run `npm start` to start the app
- The app would run on port 3000
- Access the endpoint on `localhost:3000/api/v1/`
#### Any background context you want to provide?
- This app is a node app and would require nodejs and npm installation on your local machine
- This app will require PostgreSQL Database
#### What are the relevant pivotal tracker stories?
[#163445913](https://www.pivotaltracker.com/story/show/163445913)
#### Screenshots (if appropriate)
None
#### Questions:
None

openwell committed:
#163280406- Delete Meetup Route

openwell committed:
#163280406- Delete Meetup Route 2

openwell committed:
#163280406- code refactoring

openwell committed:
#163280406- code refactoring.

openwell committed:
Chore(token, read.me): Implementation of Logic feedback

- I created a .env-sample to list the env variables and input a sample-template
- I updated my logic my getting the user-id from the token
- I updated my read.me file to give more information

[Delivers #163445913](https://www.pivotaltracker.com/story/show/163445913)

megame24 commented:
Why do you have a merge commit on this PR?

openwell commented:
I don't have a merge commit, and my develop branch can show no recent commit was made.
The last action I took was to resolve conflict and git created that commit with that name because the commit itself shows little changes when you click on changes.

megame24 commented:
> I don't have a merge commit, and my develop branch can show no recent commit was made.
> The last action I took was to resolve conflict and git created that commit with that name because the commit itself shows little changes when you click on changes.

Ok, fix your failing tests.

openwell committed:
Chore(merge conflict, hhtp code): Rebasing and update of http status code

    - I resolved merge conflicts the last PR had by doing a git rebase
    - I updated my http-status code based on feedback
    - I refactored some of my codes and added a test to the test file

    [Delivered #163504790](https://www.pivotaltracker.com/story/show/163504790)

megame24 commented on src/controller/helpers.js:
```diff
@@ -14,38 +14,55 @@ class auth {
-    auth.tokenHandler('admins', req, res, next);
+    const token = req.headers.tokens;
+    if (!token) {
+      return errorHandler(400, res, 'Token is not provided');
```
Do you think that error code 400 is appropriate for this scenario?

openwell replied:
I believe it should have been 401, authentication error

megame24 replied:
Take note and incorporate it where ever it applies.

megame24 commented on src/controller/helpers.js:
```diff
@@ -14,38 +14,55 @@ class auth {
+      req.user = { id: decoded.userId, isadmin: decoded.isAdmin };
+      return next();
+    } catch (error) {
+      return errorHandler(400, res, error);
```
In what situations will this errorHandler be called?

openwell replied:
In a case where there is server error or database error

openwell replied:
In a case where there is server error or database error

megame24 replied:
Ok, what does status code 400 mean, and do you think it applies to this case?

openwell replied:
No it should be a 500 error code

megame24 commented on src/controller/meetUpController.js:
```diff
@@ -128,11 +136,10 @@ class Controller {
+      const resp = await db.query(queries.updateVote((rows[0].votes + 1), rows[0].id));
+      const { rowCount } = await db.query(queries.votes(req.params.questionId, req.user.id));
+      return res.status(201).json({
+        status: 201,
```
When should status code 201 be returned and do you think it is appropriate to return 201 in this situation?

openwell replied:
Based on further reading i believe it should be 200

megame24 commented on src/controller/meetUpController.js:
```diff
@@ -144,11 +151,10 @@ class Controller {
+      const resp = await db.query(queries.updateVote((rows[0].votes - 1), rows[0].id));
+      const { rowCount } = await db.query(queries.votes(req.params.questionId, req.user.id));
+      return res.status(201).json({
+        status: 201,
```
When should status code 201 be returned and do you think it is appropriate to return 201 in this situation?

openwell replied:
Based on further reading i believe it should be 200

megame24 replied:
Please take note of this and incorporate it where ever it applies.

megame24 commented on src/test/allTest.js:
```diff
@@ -256,73 +250,58 @@ describe('/POST /api/v1/comments', () => {
       .send(comment2)
       .end((err, res) => {
-        res.should.have.a.status(400);
+        res.should.have.a.status(404);
```
You should be testing for more than the status code, I recommend you also test that error message is what it is meant to be. Effect this change where ever it is needed.

openwell replied:
Alright, Working on it

akinwale-habibullah commented on .evn-sample:
```diff
@@ -0,0 +1,5 @@
+DATABASE_URL=postgres://{db_username}:{db_password}@{host}:{port}/{db_name}
```
Good work. There's a little typo with your file name. Use sample.env rather than evn.sample

akinwale-habibullah replied:
Also remove the curly braces, as this can be misleading at times.

akinwale-habibullah commented on README.md:
```diff
@@ -3,55 +3,61 @@
 Clone the Repo.
 -------------
-Run npm install to install all necessary packages.
+Run `npm install` to install all necessary packages.
```
I think you should add these steps;

- Rename sample.env file
- Add the required env variables in the sample.env file
- Create database with a given name (questioner)

akinwale-habibullah commented on README.md:
```diff
@@ -3,55 +3,61 @@
-- Pull the [develop](https://github.com/openwell/questioner) branch off this repository
+- Pull the [develop](https://github.com/openwell/Questioner) branch off this repository
+- Install postgreSql 11 on your local machine or registered on any online platform
+- create a .env file
```
Why not make this rename sample.env file to .env and input all the required environment variables.

akinwale-habibullah reviewed (approved):
Kudos. I think you did a great job. Your app is quite robust. A few things you need to work on before you merge:

- In index.js please rename the error middleware you imported from errorHandler.js file
- Add some eslint rules as we discusssed.
- Return the appropriate HTTP status code when there no meetups, questions and comments records instead of 404 a user gets when you do a GET to these resources
- Return a descriptive response when admin logs in rather than http 500 status code.

akinwale-habibullah commented:
Also look at the codeclimate report, and try to work on them. I think you need to turn offf some ESLint rules.

openwell committed:
Chore(http, login etc): Implementation of feedback #2

    - I created sample.env
    - Implementation of feedback given on HTTP status code.
    - Update of ESLINT rule
    - Renaming of helpers folder to middleware
    - login Validation
    - I updated my read.me file to give more information

    [Delivers #163607712](https://www.pivotaltracker.com/story/show/163607712)
